File: python/test_engines.py
import time
import json
from parser import parse_formula
from baseline import BaselineProver
from improved import ImprovedProver
from resolution import ResolutionProver

def get_suite():
    # 16 Valid, 4 Invalid (80% / 20%)
    suite = [
        # Valid (16)
        ("A.x R(x) -> R(a)", True),
        ("R(a) -> E.x R(x)", True),
        ("~A.x R(x) -> E.x ~R(x)", True),
        ("A.x (R(x) -> S(x)) -> (A.x R(x) -> A.x S(x))", True),
        ("E.x A.y R(x,y) -> A.y E.x R(x,y)", True),
        ("A.x (R(x) -> S(x)) -> (E.x R(x) -> E.x S(x))", True),
        ("A.x (R(x) -> R(x))", True),
        ("A.x R(x) -> E.x R(x)", True),
        ("(A.x R(x) & A.x S(x)) -> A.x (R(x) & S(x))", True),
        ("A.x R(f(x)) -> E.y R(y)", True),
        ("P -> Q <-> (~Q -> ~P)", True),
        ("~~P <-> P", True),
        ("E.y A.x (P(y) -> P(x))", True),
        ("A.x (P <-> Q(x)) -> (P <-> A.x Q(x))", True),
        ("E.x E.y R(x,y) -> E.y E.x R(x,y)", True),
        ("A.x R(x) -> (A.x S(x) -> A.x (R(x) & S(x)))", True),
        # Invalid (4)
        ("E.x R(x) -> A.x R(x)", False),
        ("A.x E.y R(x,y) -> E.y A.x R(x,y)", False),
        ("R(f(a)) -> R(a)", False),
        ("E.x (R(x) -> S(x)) -> (E.x R(x) -> E.x S(x))", False)
    ]
    return suite

def run_tests():
    suite = get_suite()
    print(f"Running Evaluation on {len(suite)} problems (16 Valid, 4 Invalid)...\n")
    
    results = {}
    for name in ["Baseline", "Improved", "Resolution"]:
        results[name] = {"solved": 0, "total": len(suite), "time_ms": 0.0, "steps": 0, "max_depth": 0,
                         "inst_attempts": 0, "inst_hits": 0, "inst_wasted": 0, 
                         "clauses_gen": 0, "subsumptions": 0}

    for i, (f_str, expected) in enumerate(suite):
        try:
            f = parse_formula(f_str)
            
            # 1. Baseline
            b = BaselineProver(timeout_ms=2000)
            succ, tree, stats = b.prove(f)
            if succ == expected: results["Baseline"]["solved"] += 1
            results["Baseline"]["time_ms"] += stats.get("time_ms", 0.0)
            results["Baseline"]["steps"] += stats.get("steps", 0)
            results["Baseline"]["max_depth"] = max(results["Baseline"]["max_depth"], stats.get("max_depth", 0))
            
            # 2. Improved
            imp = ImprovedProver(timeout_ms=2000)
            succ, stats = imp.prove(f)
            if succ == expected: results["Improved"]["solved"] += 1
            results["Improved"]["time_ms"] += stats.get("time_ms", 0.0)
            results["Improved"]["steps"] += stats.get("steps", 0)
            results["Improved"]["max_depth"] = max(results["Improved"]["max_depth"], stats.get("max_depth", 0))

            # 3. Resolution
            res_prov = ResolutionProver(timeout_ms=2000)
            succ, stats = res_prov.prove(f)
            if succ == expected: results["Resolution"]["solved"] += 1
            results["Resolution"]["time_ms"] += stats.get("time_ms", 0.0)
            results["Resolution"]["steps"] += stats.get("steps", 0)
            results["Resolution"]["clauses_gen"] += stats.get("clauses_gen", 0)
            results["Resolution"]["subsumptions"] += stats.get("subsumptions", 0)
        except Exception as e:
            print(f"Error on problem {i+1}: {e}")

    print("="*110)
    print(f"{'Engine':<12} | {'Solve %':<7} | {'Time(s)':<8} | {'Steps':<6} | {'Depth':<5} | {'InstAtt':<7} | {'InstHit%':<8} | {'InstWst':<7} | {'ClsGen':<7} | {'Subsump':<7}")
    print("-" * 110)
    
    for name, r in results.items():
        solve_rate = (r["solved"] / r["total"]) * 100
        time_sec = r["time_ms"] / 1000.0
        
        inst_att = "N/A"
        inst_hit = "N/A"
        inst_wst = "N/A"
        
        cls_gen = r["clauses_gen"] if name == "Resolution" else "N/A"
        subs = r["subsumptions"] if name == "Resolution" else "N/A"
        
        print(f"{name:<12} | {solve_rate:>6.1f}% | {time_sec:>8.3f} | {r['steps']:>6} | {r['max_depth']:>5} | {inst_att:>7} | {inst_hit:>8} | {inst_wst:>7} | {cls_gen:>7} | {subs:>7}")

if __name__ == "__main__":
    run_tests()

File: test/test_engines.ml
(** test_engines.ml — Evaluation of all three engines on a 20-problem suite. *)

open Atp_ocaml.Types

let parse_formula s =
  let lexbuf = Lexing.from_string s in
  Atp_ocaml.Parser.formula_eof Atp_ocaml.Lexer.tokenize lexbuf

let suite = [
    (* Valid (16) *)
    ({|(forall x. R(x)) -> R(A)|}, true);
    ({|R(A) -> exists x. R(x)|}, true);
    ({|~(forall x. R(x)) -> exists x. ~R(x)|}, true);
    ({|(forall x. (R(x) -> S(x))) -> ((forall x. R(x)) -> (forall x. S(x)))|}, true);
    ({|(exists x. forall y. R(x,y)) -> (forall y. exists x. R(x,y))|}, true);
    ({|(forall x. (R(x) -> S(x))) -> ((exists x. R(x)) -> (exists x. S(x)))|}, true);
    ({|forall x. (R(x) -> R(x))|}, true);
    ({|(forall x. R(x)) -> (exists x. R(x))|}, true);
    ({|(forall x. R(x) /\ forall x. S(x)) -> forall x. (R(x) /\ S(x))|}, true);
    ({|(forall x. R(F(x))) -> (exists y. R(y))|}, true);
    ({|P -> Q <-> (~Q -> ~P)|}, true);
    ({|~~P <-> P|}, true);
    ({|exists y. forall x. P(y) -> P(x)|}, true);
    ({|(forall x. (P <-> Q(x))) -> (P <-> (forall x. Q(x)))|}, true);
    ({|(exists x. exists y. R(x,y)) -> (exists y. exists x. R(x,y))|}, true);
    ({|(forall x. R(x)) -> ((forall x. S(x)) -> (forall x. (R(x) /\ S(x))))|}, true);
    (* Invalid (4) *)
    ({|(exists x. R(x)) -> (forall x. R(x))|}, false);
    ({|(forall x. exists y. R(x,y)) -> (exists y. forall x. R(x,y))|}, false);
    ({|R(F(A)) -> R(A)|}, false);
    ({|(exists x. (R(x) -> S(x))) -> ((exists x. R(x)) -> (exists x. S(x)))|}, false)
]

type eval_stats = {
  mutable acc_solved : int;
  mutable acc_time_ms : float;
  mutable acc_steps : int;
  mutable acc_max_depth : int;
  mutable acc_inst_attempts : int;
  mutable acc_inst_hits : int;
  mutable acc_inst_wasted : int;
  mutable acc_clauses_gen : int;
  mutable acc_subsumptions : int;
}

let make_results () = {
  acc_solved = 0; acc_time_ms = 0.0; acc_steps = 0; acc_max_depth = 0;
  acc_inst_attempts = 0; acc_inst_hits = 0; acc_inst_wasted = 0;
  acc_clauses_gen = 0; acc_subsumptions = 0;
}

let run_eval () =
  let engines = [
    ("Baseline", (fun f -> Atp_ocaml.Baseline.prove f 2000.));
    ("Improved", (fun f -> Atp_ocaml.Improved.prove f 2000.));
    ("Resolution", (fun f -> Atp_ocaml.Resolution.prove f 2000.))
  ] in
  
  let stats_map = List.map (fun (name, _) -> (name, make_results ())) engines in
  
  Printf.printf "Running OCaml Evaluation on %d problems...\n\n" (List.length suite);
  
  List.iteri (fun _i (f_str, expected) ->
    let f = parse_formula f_str in
    List.iter (fun (name, prove_fn) ->
      let res = prove_fn f in
      let r = List.assoc name stats_map in
      if res.solved = expected then r.acc_solved <- r.acc_solved + 1;
      
      r.acc_time_ms <- r.acc_time_ms +. res.stats.time_ms;
      r.acc_steps <- r.acc_steps + res.stats.steps;
      r.acc_max_depth <- max r.acc_max_depth res.stats.max_depth;
      r.acc_inst_attempts <- r.acc_inst_attempts + res.stats.inst_attempts;
      r.acc_inst_hits <- r.acc_inst_hits + res.stats.inst_hits;
      r.acc_inst_wasted <- r.acc_inst_wasted + res.stats.inst_wasted;
      r.acc_clauses_gen <- r.acc_clauses_gen + res.stats.clauses_gen;
      r.acc_subsumptions <- r.acc_subsumptions + res.stats.subsumptions;
    ) engines
  ) suite;
  
  Printf.printf "%-12s | %-7s | %-8s | %-6s | %-5s | %-7s | %-8s | %-7s | %-7s | %-7s\n"
    "Engine" "Solve %" "Time(s)" "Steps" "Depth" "InstAtt" "InstHit%" "InstWst" "ClsGen" "Subsump";
  Printf.printf "%s\n" (String.make 110 '-');
  
  List.iter (fun (name, r) ->
    let solve_rate = (float_of_int r.acc_solved /. float_of_int (List.length suite)) *. 100. in
    let hit_rate = if r.acc_inst_attempts = 0 then 0.0 else (float_of_int r.acc_inst_hits /. float_of_int r.acc_inst_attempts) *. 100. in
    
    let inst_att = if name = "Resolution" then "N/A" else string_of_int r.acc_inst_attempts in
    let inst_hit = if name = "Resolution" then "N/A" else Printf.sprintf "%.1f%%" hit_rate in
    let inst_wst = if name = "Resolution" then "N/A" else string_of_int r.acc_inst_wasted in
    let cls_gen = if name <> "Resolution" then "N/A" else string_of_int r.acc_clauses_gen in
    let subs = if name <> "Resolution" then "N/A" else string_of_int r.acc_subsumptions in

    Printf.printf "%-12s | %6.1f%% | %8.3f | %6d | %5d | %7s | %8s | %7s | %7s | %7s\n"
      name solve_rate (r.acc_time_ms /. 1000.0) r.acc_steps r.acc_max_depth inst_att inst_hit inst_wst cls_gen subs
  ) stats_map

let () = run_eval ()
